scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java

```diff
@@ -563,26 +564,35 @@ protected void calculateKeys(byte[] masterSecret) {
 	 * concatenate a uint16 with the value N, N zero octets, a second uint16
 	 * with the value N, and the PSK itself.
 	 * 
-	 * @param psk
-	 *            the preshared key as byte array.
-	 * @return the premaster secret.
+	 * @param psk - preshared key as byte array
+	 * @param otherSecret - either is zeroes (plain PSK case) or comes 
+	 * from the EC Diffie-Hellman exchange (ECDHE_PSK), 
+	 * See <a href="http://tools.ietf.org/html/rfc4279#section-2">RFC 4279</a>:
+	 * @return
 	 */
-	protected final byte[] generatePremasterSecretFromPSK(byte[] psk) {
+	protected final byte[] generatePremasterSecretFromPSK(byte[] psk, byte[] otherSecret) {
 		/*
 		 * What we are building is the following with length fields in between:
 		 * struct { opaque other_secret<0..2^16-1>; opaque psk<0..2^16-1>; };
 		 */
-		int length = psk.length;
-
-		byte[] lengthField = new byte[2];
-		lengthField[0] = (byte) (length >> 8);
-		lengthField[1] = (byte) (length);
-
-		byte[] zero = ByteArrayUtils.padArray(new byte[0], (byte) 0x00, length);
-
-		byte[] premasterSecret = ByteArrayUtils.concatenate(lengthField, ByteArrayUtils.concatenate(zero, ByteArrayUtils.concatenate(lengthField, psk)));
-
-		return premasterSecret;
+		int pskLength = psk.length;
+
+		byte[] pskLengthField = new byte[2];
+		pskLengthField[0] = (byte) (pskLength >> 8);
+		pskLengthField[1] = (byte) (pskLength);		
+		if (otherSecret == null) {
+			byte[] zero = ByteArrayUtils.padArray(new byte[0], (byte) 0x00, pskLength);
```

Comment:
the DatagramWriter (setup up with the proper length) will do a better job.

Reply:
I added DatagramWriter instead of ByteArrayUtils.concatenate(), is this what you mean?

Reply:
I mean, that the implementation creates a lot of "short living array objects". Though the handshake is not too frequently executed, it doesn't really harm. But sometime, such "examples" are then copied into a much more frequently use scenario and then it harms.

```byte[] zero = ByteArrayUtils.padArray(new byte[0], (byte) 0x00, pskLength);```

That code first create a empty byte[] array (length 0), 
Then it creates a byte[] array of length pskLength, initialized with 0s (that's how a new array is defined in java).
Then, it fills in the 0s into the already initialized array. 
That's exactly the "bad examples", I don't like. But yes it works and doesn't harm here, because it's rarely executed ...

Reply:
yes you are right,
```java
byte[] zero = new byte[pskLength];
Arrays.fill(zero, (byte)0x00);
```
This is better than this. Am I correct?